src/WorkerContext.ts

```diff
@@ -62,7 +64,23 @@ class WorkerContext {
      * @param requestId gRPC message request id
      * @param msg gRPC message content
      */
-    log(log: rpc.IRpcLog) {
+    log(log: rpc.IRpcLog, invocationLogCtx?: InvocationLogContext): void {
+        try {
+            const logContext = new LogHookContext(log, invocationLogCtx);
+            for (const callback of worker.app.logHooks) {
+                callback(logContext);

```

Comment:
I am having a hard time grokking this. What will the callback from the logHooks do with the logContext? How will we be modifying it? Can you briefly walk me through that?

Reply:
The main idea is I will convert the log to an open telemetry format and then emit the log directly from our side. I will also tell the host (using a separate flag not included in this example) that we already emitted the log and they don't need to. Here's a simplified example:
```typescript
app.hook.log((context) => {
    otelLogger.emit({
        severityNumber: toOtelSeverityNumber(context.level),
        body: context.message,
        attributes: {
            invocationId: context.invocationContext?.invocationId,
        },
    });
});
```

In addition to emitting logs to a custom destination (like open telemetry), users could use this hook to tweak the logs that get sent to app insights:
```typescript
app.hook.log((context) => {
    if (context.category === 'system' && context.level === 'information') {
        // decrease severity level of system logs if a user finds them annoying
        context.level = 'trace';
    }

    // insert a timestamp prefix for all logs which would be more accurate than the existing timestamp property
    context.message = `${Date.now()}: ${context.message}`;
});
```

Reply:
Got it. That makes sense now. Thank you!